### ISSUE TITLE
Improve documentation for Stack.Runners

### ISSUE THREAD

**Issue by sjakobi** [commercialhaskell/stack#2694]
For https://github.com/commercialhaskell/stack/issues/2692 I had to navigate the various runners and configuration types again.

This would have been easier if the module explained e.g. that `withXConfig` functions are ordered by "power" and what information they carry in  addition to the previous type. `MiniConfig` is the "largest" type that doesn't require an existing GHC installation to construct etc.

**Comment by mgsloan** [comment]
Agreed, this is something that sometimes puzzles me, so would be a great thing to document.

One way to make things clearer might be to use a concrete monad in more places: https://github.com/commercialhaskell/stack/issues/1297
